File: regions/system/library/swift/tracing/Sources/K1s0Tracing/TraceContext.swift
import Foundation

public struct TraceContext: Sendable {
    public let traceId: String
    public let parentId: String
    public let flags: UInt8

    public init(traceId: String, parentId: String, flags: UInt8 = 1) {
        self.traceId = traceId
        self.parentId = parentId
        self.flags = flags
    }

    public func toTraceparent() -> String {
        "00-\(traceId)-\(parentId)-\(String(format: "%02x", flags))"
    }

    public static func fromTraceparent(_ s: String) -> TraceContext? {
        let parts = s.split(separator: "-").map(String.init)
        guard parts.count == 4, parts[0] == "00" else { return nil }
        guard parts[1].count == 32, parts[2].count == 16, parts[3].count == 2 else { return nil }
        guard let flags = UInt8(parts[3], radix: 16) else { return nil }
        return TraceContext(traceId: parts[1], parentId: parts[2], flags: flags)
    }
}

File: regions/system/library/dart/tracing/lib/src/trace_context.dart
class TraceContext {
  final String traceId;
  final String parentId;
  final int flags;

  const TraceContext({
    required this.traceId,
    required this.parentId,
    required this.flags,
  });

  String toTraceparent() =>
      '00-$traceId-$parentId-${flags.toRadixString(16).padLeft(2, '0')}';

  static TraceContext? fromTraceparent(String s) {
    final parts = s.split('-');
    if (parts.length != 4) return null;
    if (parts[0] != '00') return null;
    if (parts[1].length != 32) return null;
    if (parts[2].length != 16) return null;
    if (parts[3].length != 2) return null;

    final flags = int.tryParse(parts[3], radix: 16);
    if (flags == null) return null;

    return TraceContext(
      traceId: parts[1],
      parentId: parts[2],
      flags: flags,
    );
  }
}
